File: src/eih_client_test_set.cpp
#include <iostream>
#include <vector>

#include "eih_camera_api_client.h"
#define TEST_SINGLE_FUNCTION

int main() {
  std::string server_address = "172.25.181.19:15567";  // TO DO :remove

  TmEIHCamera::EIHCameraApiClient client(server_address);
  TmEIHCamera::GrpcResult grpc_result;

#ifndef TEST_SINGLE_FUNCTION
  // set capturing settings
  TmEIHConfig::SetCapturingSettingsRequest capturing_settings_req;
  capturing_settings_req.shutter_time = 36088;
  capturing_settings_req.gain = 3;
  capturing_settings_req.wb_redratio = 61;
  capturing_settings_req.wb_greenratio = 1;
  capturing_settings_req.wb_blueratio = 50;
  capturing_settings_req.focus = 6;
  capturing_settings_req.image_size = "1M";  // 1M = 1280*960, 5M = 2592*1944
  client.setCapturingSettings(grpc_result, capturing_settings_req);

  std::cout << "grpc status: " << grpc_result.status << std::endl;
  std::cout << "grpc error_message: " << grpc_result.error_message << std::endl;
  std::cout << "-------------------" << std::endl;

#else
  // Single-function
  // set shutter time
  int shutter_time = 36088;
  client.setShutterTime(grpc_result, shutter_time);
  std::cout << "grpc status: " << grpc_result.status << std::endl;
  std::cout << "grpc error_message: " << grpc_result.error_message << std::endl;
  std::cout << "-------------------" << std::endl;

  // set gain
  int gain = 4;
  client.setGain(grpc_result, gain);
  std::cout << "grpc status: " << grpc_result.status << std::endl;
  std::cout << "grpc error_message: " << grpc_result.error_message << std::endl;
  std::cout << "-------------------" << std::endl;

#endif

  return 0;
}

File: tm_eih_cam_client/eih_client_test_set.py
#!/usr/bin/env python3
"""Test script to set camera settings

This script demonstrates setting various camera settings.
"""

import sys
from tm_eih_cam_client.eih_camera_api_client import EIHCameraApiClient


def main():
    if len(sys.argv) < 2:
        print("Usage: ros2 run tm_eih_cam_client eih_client_test_set <robot_ip>")
        print("Example: ros2 run tm_eih_cam_client eih_client_test_set 192.168.10.20")
        return 1
    
    robot_ip = sys.argv[1]
    server_address = f"{robot_ip}:15567"
    
    print(f"Connecting to EIH camera at {server_address}")
    client = EIHCameraApiClient(server_address)
    
    # Test setting individual parameters
    print("\n=== Testing individual parameter settings ===")
    
    # Set focus
    print("\nSetting focus to 4...")
    result = client.set_focus(4)
    if result.status != 0:
        print(f"Failed: {result.error_message}")
    
    # Set gain
    print("Setting gain to 10...")
    result = client.set_gain(10)
    if result.status != 0:
        print(f"Failed: {result.error_message}")
    
    # Set shutter time
    print("Setting shutter time to 10000...")
    result = client.set_shutter_time(10000)
    if result.status != 0:
        print(f"Failed: {result.error_message}")
    
    # Set white balance
    print("Setting white balance (R:62, G:1, B:51)...")
    result = client.set_white_balance(62, 1, 51)
    if result.status != 0:
        print(f"Failed: {result.error_message}")
    
    # Test setting all parameters at once
    print("\n=== Testing batch parameter setting ===")
    print("Setting all capturing parameters...")
    result = client.set_capturing_settings(
        shutter_time=36088,
        gain=4,
        wb_redratio=62,
        wb_greenratio=1,
        wb_blueratio=51,
        focus=6,
        image_size="1M"
    )
    
    if result.status == 0:
        print("Successfully set all parameters")
    else:
        print(f"Failed to set parameters: {result.error_message}")
    
    # Verify settings
    print("\n=== Verifying settings ===")
    result, settings = client.get_capturing_settings()
    if result.status == 0 and settings:
        print(f"Current settings:")
        print(f"  Shutter time: {settings['shutter_time']['current_value']}")
        print(f"  Gain: {settings['gain']['current_value']}")
        print(f"  Focus: {settings['focus']['current_value']}")
        print(f"  Image size: {settings['image_size']}")
        print(f"  WB Red: {settings['white_balance']['red_ratio']['current_value']}")
        print(f"  WB Green: {settings['white_balance']['green_ratio']['current_value']}")
        print(f"  WB Blue: {settings['white_balance']['blue_ratio']['current_value']}")
    
    print("\n=== Test completed ===")
    return 0


if __name__ == '__main__':
    sys.exit(main())
